firedrake/functionspacedata.py

```diff
@@ -454,134 +453,71 @@ def boundary_nodes(self, V, sub_domain, method):
             key = (entity_dofs_key(V.finat_element.entity_dofs()), sdkey, method)
             return get_boundary_nodes(V.mesh(), key, V)
 
-    def get_map(self, V, entity_set, map_arity, bcs, name, offset, parent,
-                kind=None):
+    def lgmap(self, V, bcs, lgmap=None):
+        assert len(V) == 1, "lgmap should not be called on MixedFunctionSpace"
+        V = V.topological
+        if lgmap is None:
+            lgmap = V.dof_dset.lgmap
+        if bcs is None or len(bcs) == 0:
+            return lgmap
+        for bc in bcs:
+            fs = bc.function_space()
+            # Unwind proxies for ComponentFunctionSpace, but not
+            # IndexedFunctionSpace.
+            while fs.component is not None and fs.parent is not None:
+                fs = fs.parent
+            if fs.topological != V:
+                raise RuntimeError("DirichletBC defined on a different FunctionSpace!")
+        # Boundary condition list *must* be collectively ordered already.
+        # Key is a sorted list of bc subdomain, bc method, bc component.
+        bc_key = []
+        for bc in bcs:
+            bc_key.append(bc.domain_args + (bc.method, ) + (bc.function_space().component, ))
+        bc_key = tuple(sorted(bc_key))
+        unblocked = any(bc.function_space().component is not None for bc in bcs)
+        if unblocked:
+            indices = lgmap.indices.copy()
+            bsize = 1
+        else:
+            indices = lgmap.block_indices.copy()
+            bsize = lgmap.getBlockSize()
+        nodes = []
+        for bc in bcs:
+            if bc.function_space().component is not None:
+                nodes.append(bc.nodes * V.value_size + bc.function_space().component)
+            elif unblocked:
+                tmp = bc.nodes * V.value_size
+                for i in range(V.value_size):
+                    nodes.append(tmp + i)
+            else:
+                nodes.append(bc.nodes)
+        nodes = numpy.unique(numpy.concatenate(nodes))
+        indices[nodes] = -1
+        return PETSc.LGMap().create(indices, bsize=bsize, comm=lgmap.comm)
```

Comment:
It's so simple, we just unroll the indices for the L2G map and set the boundary ones negative. This works transparently in extruded and everywhere else because bc.nodes reports the boundary nodes for all nodes in the dof set.